src/main/java/com/example/onlinebookstore/service/order/OrderServiceImpl.java

```diff
@@ -0,0 +1,102 @@
+    private final OrderItemMapper orderItemMapper;
+
+    @Override
+    public OrderResponseDto placeOrder(OrderRequestDto orderRequestDto, User user) {
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseThrow(() -> new EntityNotFoundException(
+                        String.format("Can't find shopping cart for user with id: %d",
+                                user.getId())));
+
+        BigDecimal total = shoppingCart.getCartItems().stream()
+                .map(cartItem -> cartItem.getBook().getPrice()
+                        .multiply(BigDecimal.valueOf(cartItem.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add);
+
+        Order order = new Order();
+        order.setUser(user);
+        order.setStatus(Status.PENDING);
+        order.setOrderDate(LocalDateTime.now());
+        order.setShippingAddress(orderRequestDto.getShippingAddress());
+        order.setTotal(total);
+        Order savedOrder = orderRepository.save(order);
+        Set<OrderItem> orderItems = createOrderItems(savedOrder, shoppingCart);
+        order.setOrderItems(orderItems);
+
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    @Override
+    public List<OrderResponseDto> getOrderHistory(Long userId) {
+        List<Order> orders = orderRepository.findByUserId(userId);
+
+        return orders.stream()
+                .map(orderMapper::toDto)
+                .toList();
+    }
+
+    @Override
+    public List<OrderItemResponseDto> getOrderItems(Long orderId) {
+        List<OrderItem> orderItems = orderItemRepository.findByOrderId(orderId);
+        return orderItems.stream().map(orderItemMapper::toDto).toList();
+    }
+
+    @Override
+    public OrderItemResponseDto getOrderItem(Long orderId, Long itemId) {
+        OrderItem orderItem = orderItemRepository.findById(itemId)
+                .orElseThrow(() -> new RuntimeException("Order item not found"));
+        if (!orderItem.getOrder().getId().equals(orderId)) {
+            throw new RuntimeException("Order item does not belong to the specified order");
+        }
+        return orderItemMapper.toDto(orderItem);
+    }
+
+    private Set<OrderItem> createOrderItems(Order order, ShoppingCart shoppingCart) {
+        Set<OrderItem> orderItems = shoppingCart.getCartItems().stream()
+                .map(item -> createOrderItem(item, order))
+                .collect(Collectors.toSet());
+        orderItemRepository.saveAll(orderItems);
+        return orderItems;
+    }
+
+    private OrderItem createOrderItem(CartItem item, Order order) {
+        OrderItem orderItem = new OrderItem();
+        orderItem.setOrder(order);
+        orderItem.setBook(item.getBook());
+        orderItem.setQuantity(item.getQuantity());
+        orderItem.setPrice(item.getBook().getPrice()
+                .multiply(BigDecimal.valueOf(item.getQuantity())));
+        return orderItem;
+    }
```

Comment:
rename those methods to something like `mapCartItemsToOrderItems`, overall I even might have created a mapper class for it, to not keep it in the service, because that is not the responsibility of the service TBH